feat(cli): add support for custom configurations formats in JSON, JSON5, JS, YAML, or from a shared module

peterkc opened the pull request:
Add support for custom configurations from any of the following:

- a path specified with `--config`
- a config file relative to `process.cwd()`
- a path from an environment variable `FORMAT_PACKAGE_CONFIG`

Supported custom configuration formats: JSON, JSON5, JS, and YAML.

peterkc commented on .circleci/config.yml:
```diff
@@ -4,7 +4,7 @@ aliases:
     working_directory: *working-directory
     docker:
-      - image: 'circleci/node:9.10'
+      - image: 'circleci/node:8.16'
```
This is to resolve circleci setup, `get-caller-file` requiring`package.json/engines/node` to be `"node": "6.* || 8.* || >= 10.*"`.

peterkc commented on __tests__/fixtures/index.spec.js:
```diff
@@ -1,6 +1,9 @@
 const path = require('path');
 
+// eslint-disable-next-line
+const stripAnsi = require('strip-ansi');
```
`jest` snapshots when calling update in a shell with tty will include ansi characters. By striping ansi, this will insure that tty/non-tty will work as expected for all `expect`.

camacho replied:
Let's consider using a [custom serializer](https://github.com/joaogranado/jest-serializer-ansi) for the snapshots

peterkc replied:
@camacho Done.

peterkc commented on lib/cli/__snapshots__/parse.spec.js.snap:
```diff
@@ -17,7 +17,6 @@ Object {
     "alias": "config",
     "demandOption": false,
     "describe": "Location of a config file",
-    "transform": [Function],
```
Replacing with `coerce` to resolve paths with `require.resolve`.

peterkc commented on lib/cli/parse.js:
```diff
@@ -39,6 +37,12 @@ const parser = require('yargs')('run --help')
+    return require.resolve(configPath, {
+      paths: [process.cwd()],
+    });
+  })
```
Use `require.resolve` instead of `path.resolve` to support configs in node_modules.

camacho replied:
Are there any drawbacks for this approach that would make it not backward compatible?

camacho replied:
After running a few tests, it appears that this will not work for resolving local config files. For example, in this repository, just trying to load the `package.json` file using the `require.resolve` logic above results in an error:

```js
require.resolve("package.json", {
  paths: [process.cwd()],
});
// Thrown: Error: Cannot find module 'package.json'
```

peterkc replied:
@camacho I have tried running it with `node lib/cli/index.js package.json` and did not get an error. Can you provide additional info?

camacho replied:
Ah, I think what we are seeing is a change in behavior. The previous implementation did not require a `./` notation for local config paths, allowing a user to write `format-package -w --config format-package-config.js`. This change breaks that previous behavior by requiring any config path to starts with ".".

Other configurations for CLIs, as I remember, do not require the config argument to start with a "./" for local files (such as Jest/ESLint/etc). I do worry that we are going against what I would first expect as a new users (although that's just from my perspective). 

Currently, it is possible to create a config file that imports a node module and re-exports to use shared logic, without this change. But supporting the node module directly from the command line is a nice touch as well. Is there a smart way we can support both, with priority given to finding a local file and falling back to looking for a dependency? 

We could also always look for a config file (which yargs makes pretty easy) and allow for an `extends` property or something of the sort like ESLint/TS/Babel/etc.?

I don't know what the right answer is, but I think continuing to support relative config paths that don't start with "./" is meaningful.

camacho replied:
We could also adopt something like [`cosmiconfig`](https://github.com/davidtheclark/cosmiconfig) so a config file is always looked for automatically and the config file could then export a shared config dependency (like ESLint/Typescript/Babel/etc.). 

This would leave the CLI `-c` option as a relative path to a different config file location.

peterkc replied:
@camacho I like the idea of using `cosmiconfig`. It's highly configurable and I can move the support for using a `node_module` to that. Will update the PR when it is done and all tests are passing. 

When `--config my-format.js` is provided with a file path that exists.

```
[ '/my-project/my-format.js',
  'format-package.js',
  'format-package.yaml',
  'format-package.yml',
  'format-package.json',
  'format-package.config.js',
  'format-package.config.yaml',
  'format-package.config.yml',
  'package.json' ]
```

When `--config my-format` is provided with a module name that exists.

```
[ '/my-project/node_modules/my-format/index.js',
  'format-package.js',
  'format-package.yaml',
  'format-package.yml',
  'format-package.json',
  'format-package.config.js',
  'format-package.config.yaml',
  'format-package.config.yml',
  'package.json' ]
```

Else search for one of these.

```
[ 'format-package.js',
  'format-package.yaml',
  'format-package.yml',
  'format-package.json',
  'format-package.config.js',
  'format-package.config.yaml',
  'format-package.config.yml',
  'package.json' ]
```

If no config is found, return `lib/defaults/index.js.

peterkc replied:
In regards to the extends, I'll make that as a stretch or another PR.

peterkc replied:
@camacho `cosmiconfig` added.

camacho replied:
@peterkc this is feels like a great approach to finding the config and feels intuitive 👍

peterkc commented on lib/cli/parse.js:
```diff
@@ -9,13 +7,13 @@ const parser = require('yargs')('run --help')
       alias: 'globs',
     });
   })
+  .env('FORMAT')
```
All for using environment variables for options.

camacho replied:
Since this environment variable lives in a global scope, can we choose a more specific name such as `PACKAGE_FORMAT_CONFIG`?

peterkc replied:
@camacho Done.

peterkc commented on lib/defaults/transformations.spec.js:
```diff
@@ -13,7 +16,9 @@ describe('default transformations', () => {
-      expect(transformations.scripts('key', 'value')).toMatchSnapshot();
+      expect(
+        stripAnsi(transformations.scripts('key', 'value'))
+      ).toMatchSnapshot();
```
Format is a result of running `yarn format`?

peterkc commented on package.json:
```diff
@@ -32,7 +32,9 @@
-    "test": "jest"
+    "test": "jest",
+    "test:coverage": "jest --coverage",
+    "test:update-snapshot": "jest --updateSnapshot --color=false"
```
Useful run scripts for dev.

camacho replied:
these are unnecessary since they can be accomplished by running `yarn test --coverage` and `yarn test -u`

peterkc replied:
@camacho Removed.

camacho commented on lib/cli/error.spec.js:
```diff
@@ -22,25 +25,25 @@ describe('error', () => {
   it('logs generic message and exists with 1 when called without error', () => {
     logErrorAndExit();
-    expect(console.error.mock.calls[0][0]).toMatchSnapshot();
+    expect(stripAnsi(console.error.mock.calls[0][0])).toMatchSnapshot();
```
We can use a [custom serializer](https://github.com/joaogranado/jest-serializer-ansi) here

camacho commented on lib/cli/index.js:
```diff
@@ -55,6 +55,7 @@ async function execute(argv) {
       files.map(file => path.resolve(file)).map(handleFile(options, config))
     );
 
+    /* istanbul ignore next */
```
👍

camacho commented:
This is a great addition @peterkc! 

📦The ability to use a dependency for a configuration is fantastic 
💻Adding a way to pass configs through environment variables is a nice addition
📖I really appreciate you taking the time to update the `README` for everyone else

Three quick points of feedback:
1. Let's pick a more specific environment variable name scope like `PACKAGE_FORMAT`
1. Instead of manually (and rightly) stripping ASCII for each Jest snapshot, we should use a [custom serializer](https://github.com/joaogranado/jest-serializer-ansi)
1. I'd like to leave the `package.json` scripts as few as possible and the additions in this PR do not enable any functionality that was not already possible with the already defined scripts

peterkc commented:
@camacho Have update the PR with the changes requested. 

- Rebased to latest changes in master.
- Added yarn.lock to .gittattributes to avoid dealing with merges, yarn will automatically resolve conflicts.

camacho commented on .gitattributes:
```diff
@@ -0,0 +1,3 @@
+# Don't merge changes to these generated files, because the result
+# may be invalid.
+yarn.lock                    merge=binary
```
👍

camacho commented on lib/cli/parse.js:
```diff
@@ -39,6 +37,12 @@ const parser = require('yargs')('run --help')
       type: 'array',
     },
   })
+  .coerce('config', configPath => {
```
small nit - could we move this new logic to where the `options.c.transform` function previously was to keep all the option configurations grouped together?

peterkc replied:
@camacho This is deprecated by `lib/cli/config.js` and `cosmiconfig`.

peterkc commented on lib/cli/index.spec.js:
```diff
@@ -1,137 +1,58 @@
+    expect(logErrorAndExit).toHaveBeenCalled();
+    expect(console.warn).toHaveBeenCalledTimes(1);
   });
 });
```
The addition of `cosmiconfig` simplified `index.spec.js` by delegating all config parsing tests to `config.spec.js`.

peterkc commented on package.json:
```diff
@@ -29,7 +29,7 @@
   "scripts": {
     "docs": "scripts/markdown.js",
     "format": "npm-run-all format:*",
-    "format:pkg": "./lib/cli/index.js -wp",
```
`-p` was an unknown option causing the positional arguments to be ignored when provided.

peterkc commented on package.json:
```diff
@@ -29,7 +29,7 @@
     "docs": "scripts/markdown.js",
     "format": "npm-run-all format:*",
-    "format:pkg": "./lib/cli/index.js -wp",
+    "format:pkg": "./lib/cli/index.js -w package.json",
```
Need to format only the root `package.json` and ignore the ones in `examples/*`.

peterkc commented on package.json:
```diff
@@ -52,10 +52,21 @@
+    "cosmiconfig": "^5.2.1",
+    "fs-extra": "^8.1.0",
     "globby": "10.0.1",
+    "resolve-from": "^5.0.0",
```
A `sindresorhus` module like require.resolve() but from a given path and will return undefined instead of throwing.

peterkc commented on package.json:
```diff
@@ -52,10 +52,21 @@
     "chalk": "^2.4.2",
     "convert-hrtime": "^3.0.0",
+    "cosmiconfig": "^5.2.1",
+    "fs-extra": "^8.1.0",
```
`fs-extra` is a drop in replacement for `fs` with promises and use of `graceful-fs` to prevent EMFILE errors.

camacho commented on lib/cli/parse.js:
```diff
@@ -39,6 +37,7 @@ const parser = require('yargs')('run --help')
       type: 'array',
     },
   })
+  .strict()
```
This is a nice addition. I had actually forgotten the API and had been using `-q` for "quiet" for quite a while now 🤦‍♂

camacho commented on lib/cli/config.js:
```diff
@@ -0,0 +1,114 @@
+    .unique(),
+  transformations: Joi.func().optional(),
+  formatter: Joi.func().optional(),
+});
```
Nice to see the configuration getting some stronger validation logic 👍

camacho commented on README.md:
````diff
@@ -354,20 +344,108 @@ The CLI accepts a series of files or globs to be formatted, as well as a set of
+```
+
+### with a config file
+
````
Can we add some explanation here that the package will automatically look for these files, in a numbered order, if no `--config` option is passed? Maybe something like:

If no `--config` option is provided, the package will automatically look for a defined configuration, proceeding down an ordered lookup list until one is found:
1. `format-package` property in `package.json`
1. `format-package.json`
1. `format-package.js`
1. `format-package.yaml`

Each type of config file is expected to export a config object in the language denoted by their file extension.

camacho commented on lib/cli/index.js:
```diff
@@ -30,22 +36,14 @@ const handleFile = ({ write, verbose }, config) => async filePath => {
+    const { globs, config: configPath, ignore, ...options } = parser(argv);
 
-    const config = configPath ? require(configPath) : {};
+    const { config } = await configSearch.search({ configPath });
```
We should likely be handling the possible `error` property value being set from this call

peterkc replied:
@camacho the error is handled by the `.catch` in the `search()` in `config.js`. The current behavior is to fallback to the default configuration in `lib/defaults/index.js`. In a different PR I would like to rework the way how we do logging (e.g `--loglevel warn`).

camacho replied:
The second part is more what I was curious about. Improving logging is a good idea.

camacho commented on README.md:
````diff
@@ -354,20 +344,108 @@ The CLI accepts a series of files or globs to be formatted, as well as a set of
+```
+
+### with an environment variable `FORMAT_PACKAGE_CONFIG`
+
````
A small bit of text (1 sentence) here explaining what is happening would be nice for any developers who are not be super familiar with command line tools and could be unaware of how an environment variable works.

camacho commented on README.md:
````diff
@@ -354,20 +344,108 @@ The CLI accepts a series of files or globs to be formatted, as well as a set of
 yarn format-package --help
 ```
 
+## Config
````
I'm not sure where it should go, but I think we should probably add a small blurb letting developers know that the configuration can be partial.

camacho commented:
@peterkc a few small pieces of feedback and 1 question about handling errors. Overall, this looks great and your work will make configuration much, much better 👍

peterkc commented on README.md:
````diff
@@ -354,20 +344,108 @@ The CLI accepts a series of files or globs to be formatted, as well as a set of
+yarn format-package -c @acme/format-config package.json
+yarn format-package -c ./configs/format-config.js package.json
+yarn format-package -c format-config.js package.json
+```
````
OK. Let me see if I can clean that up a bit.

peterkc commented:
@camacho Updated the docs in regards to configuration orders and examples. Take a look when you get a chance.

camacho commented:
@peterkc I think you should have merge privileges now that you're a collaborator. If not, I can merge this.
